tikv-client/src/main/java/com/pingcap/tikv/TiBatchWrite2PCClient.java

```diff
@@ -0,0 +1,584 @@
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+
+      ByteWrapper that = (ByteWrapper) o;
+
+      return Arrays.equals(bytes, that.bytes);
+    }
+
+    @Override
+    public int hashCode() {
+      return Arrays.hashCode(bytes);
+    }
+  }
+
+  public static class BytePairWrapper {
+    private byte[] key;
+    private byte[] value;
+
+    public BytePairWrapper(byte[] key, byte[] value) {
+      this.key = key;
+      this.value = value;
+    }
+
+    public byte[] getKey() {
+      return key;
+    }
+
+    public byte[] getValue() {
+      return value;
+    }
+  }
+
+  /** buffer spark rdd iterator data into memory */
+  private static final int WRITE_BUFFER_SIZE = 1024 * 1024;
+
+  /**
+   * TiKV recommends each RPC packet should be less than ~1MB. We keep each packet's Key+Value size
+   * below 16KB.
+   */
+  private static final int TXN_COMMIT_BATCH_SIZE = 16 * 1024;
+
+  /** unit is second */
+  private static final long DEFAULT_BATCH_WRITE_LOCK_TTL = 3000;
+
+  private static final Logger LOG = LoggerFactory.getLogger(TiBatchWrite2PCClient.class);
+
+  private TxnKVClient kvClient;
+  private RegionManager regionManager;
+
+  /** start timestamp of transaction which get from PD */
+  private final long startTs;
+
+  public TiBatchWrite2PCClient(TxnKVClient kvClient, long startTime) {
+    this.kvClient = kvClient;
+    this.regionManager = kvClient.getRegionManager();
+    this.startTs = startTime;
+  }
+
+  /**
+   * 2pc - prewrite primary key
+   *
+   * @param backOffer
+   * @param primaryKey
+   * @param value
+   * @return
+   */
+  public void prewritePrimaryKey(BackOffer backOffer, byte[] primaryKey, byte[] value)
+      throws TiBatchWriteException {
+    String error = this.doPrewritePrimaryKey(backOffer, primaryKey, value);
+    if (error != null) {
+      throw new TiBatchWriteException("prewrite primary key error: " + error);
+    }
+  }
+
+  private String doPrewritePrimaryKey(BackOffer backOffer, byte[] key, byte[] value) {
+    TiRegion tiRegion = this.regionManager.getRegionByKey(ByteString.copyFrom(key));
+    BatchKeys batchKeys = new BatchKeys(tiRegion.getId(), Arrays.asList(value));
+
+    Kvrpcpb.Mutation mutation =
+        Kvrpcpb.Mutation.newBuilder()
+            .setKey(ByteString.copyFrom(key))
+            .setValue(ByteString.copyFrom(value))
+            .setOp(Kvrpcpb.Op.Put)
+            .build();
+    List<Kvrpcpb.Mutation> mutationList = Collections.singletonList(mutation);
+
+    // send rpc request to tikv server
+    long regionId = batchKeys.getRegioId();
+    long lockTTL = getTxnLockTTL(this.startTs);
+    ClientRPCResult prewriteResult =
+        this.kvClient.prewrite(backOffer, mutationList, key, lockTTL, this.startTs, regionId);
+    if (!prewriteResult.isSuccess() && !prewriteResult.isRetry()) {
+      return prewriteResult.getError();
+    }
+    if (prewriteResult.isRetry()) {
```

Comment:
I am kinda confused with this logic. AbstractGRPCClient has a callWithRetry method. If it is a KV error, we have a KVErrorHandler. If it's a PD error, we have a PDErrorHandler.

Reply:
what if primary key's region changes? should we retry from getting the primary key's new region?

Reply:
@birdstorm what do u think?

Reply:
yes, we can just retry here. region[old]'s requests will be all retried on region[new]